include/directx/d3dx12.h

```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        if (result == E_INVALIDARG) {
+            // Shader model may not be supported
+            m_dShaderModel.HighestShaderModel = (D3D_SHADER_MODEL)0;
+            result = S_OK; // Continue initialization
+        }
+
+        return result;
+    }
+
+    // Helper function to decide the highest root signature supported
+    // Must be updated whenever a new root signature version is added to the d3d12.h header
+    HRESULT QueryHighestRootSignatureVersion()
+    {
+        HRESULT result;
+
+        // ROOT_SIGNATURE_VERSION_1_X
+        for (int i = D3D_ROOT_SIGNATURE_VERSION_1_1; i >= D3D_ROOT_SIGNATURE_VERSION_1; i--) {
+            m_dRootSignature.HighestVersion = static_cast<D3D_ROOT_SIGNATURE_VERSION>(i);
+            result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_ROOT_SIGNATURE, &m_dRootSignature, sizeof(D3D12_FEATURE_DATA_ROOT_SIGNATURE));
+            if (SUCCEEDED(result) || result != E_INVALIDARG) {
+                return result;
+            }
+        }
+
+        // No version left. Set to invalid value and continue.
+        m_dRootSignature.HighestVersion = (D3D_ROOT_SIGNATURE_VERSION)0;
+        return S_OK;
+    }
+
+    // Helper funcion to decide the highest feature level
+    HRESULT QueryHighestFeatureLevel()
+    {
+        HRESULT result;
+
+        // Check against a list of all feature levels present in d3dcommon.h
+        // Needs to be updated for future feature levels
+        D3D_FEATURE_LEVEL allLevels[] = {
+            D3D_FEATURE_LEVEL_12_2,
+            D3D_FEATURE_LEVEL_12_1,
+            D3D_FEATURE_LEVEL_12_0,
+            D3D_FEATURE_LEVEL_11_1,
+            D3D_FEATURE_LEVEL_11_0,
+            D3D_FEATURE_LEVEL_10_1,
+            D3D_FEATURE_LEVEL_10_0,
+            D3D_FEATURE_LEVEL_9_3,
+            D3D_FEATURE_LEVEL_9_2,
+            D3D_FEATURE_LEVEL_9_1,
+            D3D_FEATURE_LEVEL_1_0_CORE
+        };
+
+        D3D12_FEATURE_DATA_FEATURE_LEVELS dFeatureLevel;
+        dFeatureLevel.NumFeatureLevels = sizeof(allLevels) / sizeof(D3D_FEATURE_LEVEL);
+        dFeatureLevel.pFeatureLevelsRequested = allLevels;
+
+        result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_FEATURE_LEVELS, &dFeatureLevel, sizeof(D3D12_FEATURE_DATA_FEATURE_LEVELS));
+        if (SUCCEEDED(result)) {
+            m_eMaxFeatureLevel = dFeatureLevel.MaxSupportedFeatureLevel;
+        } else if (result == DXGI_ERROR_UNSUPPORTED) {
+            m_eMaxFeatureLevel = (D3D_FEATURE_LEVEL)0;
+            result = S_OK;
+        }
+        return result;
+    }
+
+private: // Member data
+    // Pointer to the underlying device
+    ID3D12Device* m_pDevice;
+
+    // Stores the error code from initialization
+    HRESULT m_hStatus;
+
+    // Feature support data structs
+    D3D12_FEATURE_DATA_D3D12_OPTIONS m_dOptions;
+    // std::vector<D3D12_FEATURE_DATA_ARCHITECTURE> m_dArchitecture; // Cat2 NodeIndex, should be deprecated

```

Comment:
I'd suggest removing these commented out entries.

Reply:
I left them in there to remind me the entries in the feature enum. I can remove them now that the implementation is done.

Reply:
Actually there are several places where I created a feature data struct in the function instead of using these commented out member struct. I guess it's better to use these member structs to avoid repeated allocation? See FormatSupport for an example.

Reply:
I actually think it's better to create these just where you need them.  That will be very cheap / free stack allocation, only lasting as long as the function that uses it, and makes clear that the usage of the object is only temporary.

Reply:
Resolved. Used temporary data structs and removed unused member structs.